### PR TITLE
nested aux-build in tests/rustdoc/ tests

### PR DESCRIPTION
**Pull request by EtomicBomb**
* Fixes bug that prevented using nested aux-build in `tests/rustdoc/` tests. Before, `fn document` and the auxiliary builder disagreed about where to find the nested aux-build source file (`auxiliary/auxiliary/aux.rs` vs `auxiliary/aux.rs`), preventing them from building. Picked the latter in line with other builders in compiletest.
* Adds `//@ doc-flags` header, which forwards flags to rustdoc and not rustc. 
* Adds `//@ unique-doc-out-dir` header, which sets the --out-dir for the rustdoc invocation to a unique directory: `<root out dir>/docs/<test name>/doc`
* Changes working directory of the rustdoc invocation to the root out directory (common among all aux-builds). Prior art: exec_compiled_test in runtest.rs
* Adds tests that use nested aux builds and new headers

These changes provide useful capabilities for writing rustdoc tests on their own. They are also needed to test the implementation for the [mergable-rustdoc-cross-crate-info](https://github.com/rust-lang/rfcs/pull/3662) RFC.

try-job: x86_64-msvc

### CODE REVIEW

**Comment by compiler-errors** [comment]
This PR's history should be cleaned quite a lot, at least rebasing out all of the merge commits but also probably the useless commits that represent intermediate work states like "code from this morning".

r? rustdoc

**Comment by EtomicBomb** [comment]
Rebased the history! This is my first PR, so I'm just learning the ropes!

**Review by notriddle — COMMENTED** [review]
AFAICT, this seems fine. In particular, I love these test cases!

I'm not as familiar with the deep guts of compiletest, though.

CC @jieyouxu

**Comment by jieyouxu** [comment]
> I'm not as familiar with the deep guts of compiletest, though.
> 
> CC @jieyouxu

I'll try to take a look tomorrow for the compiletest changes.

File: src/tools/compiletest/src/runtest.rs
Comment (jieyouxu) [inline src/tools/compiletest/src/runtest.rs:1596]: Question: do rustdoc tests support revisions like ui tests, if so, would this be a problem with revisioned rustdoc tests?

**Review by jieyouxu — APPROVED** [review]
Sorry for the wait, I just got to this PR today. The compiletest changes LGTM for the most part, just a question about potential test revision interactions.

**Comment by EtomicBomb** [comment]
I just rebased to fix some merge conflicts. I think the PR still has an extraneous S-waiting-on-author label that rustbot assigned before I removed the merge commits last week.

**Comment by EtomicBomb** [comment]
I just rebased because of https://github.com/rust-lang/rust/pull/125443, which I think means it needs another review!

File: src/tools/compiletest/src/runtest.rs
Comment (jieyouxu) [reply src/tools/compiletest/src/runtest.rs:1588]: > I’ve never seen them used, and htmldocck itself doesn’t know about them, so I guess not.

Actually it's very easy:

https://github.com/rust-lang/rust/blob/1ddedbaa5919b7b3e70d984660e21e844c615c97/src/tools/compiletest/src/runtest.rs#L2661-L2662

no revisions are not supported for `tests/rustdoc` 😄

**Review by jieyouxu — APPROVED** [review]
Thanks, r=me for the compiletest changes.

**Pull request merged**
